openedx/features/enterprise_support/middleware.py

```diff
@@ -28,3 +34,33 @@ def process_request(self, request):
         """
         if 'enterprise_customer' not in request.session and request.user.is_authenticated:
             request.session['enterprise_customer'] = api.enterprise_customer_for_request(request)
+
+
+def set_learner_active_enterprise(user=None, backend=None, *args, **kwargs):
+    """
+    Make 'active' a user's enterprise,
+    if the currently 'active' enterprise in EnterpriseCustomerUser does not match the SAML idp-enterprise
+    """
+    if is_provider_saml(backend.name, kwargs):
+        request = kwargs["strategy"].request
+        idp_name = saml_idp_name(backend.name, kwargs['response']['idp_name'])
+        session_idp = get_enterprise_customer_from_session(request)
+        if not session_idp or idp_name != session_idp['identity_provider']:
+            enterprise_learner_data = api.get_enterprise_learner_data(user)
+            enterprise_learner = [enterprise_learner for enterprise_learner in enterprise_learner_data if
+                                      enterprise_learner['enterprise_customer']['identity_provider'] == idp_name]
+            if enterprise_learner:
+                uuid = enterprise_learner[0]['enterprise_customer']['uuid']
+                try:
+                    api.EnterpriseApiClient(user=user).post_active_enterprise_customer(user.username, uuid, True)
+                    activate_learner_enterprise(request, enterprise_learner[0]['enterprise_customer'])
+                except api.EnterpriseApiException:
```

Comment:
Don't need to catch exception here since we are catching it upstream.

Reply:
@ziafazal I have addressed the changes. Please take a look.